packages/flutter/lib/src/widgets/widget_inspector.dart

```diff
@@ -0,0 +1,526 @@
+    );
+
+    if (state != _lastState) {
+      _lastState = state;
+      _picture = _buildPicture(state);
+    }
+    builder.addPicture(layerOffset, _picture);
+  }
+
+  static ui.Picture _buildPicture(_InspectorOverlayRenderState state) {
+    final ui.PictureRecorder recorder = new ui.PictureRecorder();
+    final Canvas canvas = new Canvas(recorder, state.overlayRect);
+    final Size size = state.overlayRect.size;
+
+    final Paint fillPaint = new Paint()
+      ..style = PaintingStyle.fill
+      ..color = _kHighlightedRenderObjectFillColor;
+
+    final Paint borderPaint = new Paint()
+      ..style = PaintingStyle.stroke
+      ..strokeWidth = 1.0
+      ..color = _kHighlightedRenderObjectBorderColor;
+
+    // Highlight the selected renderObject.
+    final Rect selectedPaintRect = state.selected.rect.deflate(0.5);
+    canvas
+      ..save()
+      ..transform(state.selected.transform.storage)
+      ..drawRect(selectedPaintRect, fillPaint)
+      ..drawRect(selectedPaintRect, borderPaint)
+      ..restore();
+
+    // Show all other candidate possibly selected elements. This helps selecting
+    // render objects by selecting the edge of the bounding box shows all
+    // elements the user could toggle the selection between.
+    for (_TransformedRect transformedRect in state.candidates) {
+      canvas
+        ..save()
+        ..transform(transformedRect.transform.storage)
+        ..drawRect(transformedRect.rect.deflate(0.5), borderPaint)
+        ..restore();
+    }
+
+    final Rect targetRect = MatrixUtils.transformRect(
+        state.selected.transform, state.selected.rect);
+    final Offset target = new Offset(targetRect.left, targetRect.center.dy);
+    final double offsetFromWidget = 9.0;
+    final double verticalOffset = (targetRect.height) / 2 + offsetFromWidget;
+
+    _paintDescription(
+        canvas, state.tooltip, target, verticalOffset, size, targetRect);
+
+    // TODO(jacobr): provide an option to perform a debug paint of just the
+    // selected widget.
+    return recorder.endRecording();
+  }
+}
+
+const double _kScreenEdgeMargin = 10.0;
+const double _kTooltipPadding = 5.0;
+const double _kInspectButtonMargin = 10.0;
+
+/// Interpret pointer up events within with this margin as indicating the
+/// pointer is moving off the device.
+const double _kOffScreenMargin = 1.0;
+
+const TextStyle _messageStyle = const TextStyle(
+  color: const Color(0xFFFFFFFF),
+  fontSize: 10.0,
+  height: 1.2,
+);
+
+final int _kMaxTooltipLines = 5;
+final Color _kTooltipBackgroundColor = const Color.fromARGB(230, 60, 60, 60);
+final Color _kHighlightedRenderObjectFillColor = const Color.fromARGB(128, 128, 128, 255);
+final Color _kHighlightedRenderObjectBorderColor = const Color.fromARGB(128, 64, 64, 128);
+
+void _paintDescription(Canvas canvas, String message, Offset target,
+    double verticalOffset, Size size, Rect targetRect) {
+  canvas.save();
+  final TextPainter textPainter = new TextPainter()
```

Comment:
Done. The text painter is now cached as it really only changes if the device is rotated or the selected RenderObject changes.